plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java

```diff
@@ -35,121 +35,189 @@
+                if (!removedPools.isEmpty()) {
+                    for (String uuid : removedPools) {
+                        removeStoragePool(uuid);
                     }
+                }
+            }
+        }
 
+        if (rbdstoragePool != null && !rbdstoragePool.isEmpty()) {
+            synchronized (rbdstoragePool) {
+                Set<String> removedPools = new HashSet<>();
+                for (String uuid : rbdstoragePool.keySet()) {
+                    RbdStoragePool rbdStoragePool = rbdstoragePool.get(uuid);
+                    runHeartbeatToPool(nfsStoragePool, rbdStoragePool, uuid, removedPools);
+                    continue;
                 }
 
-                String result = null;
-                for (int i = 1; i <= _heartBeatUpdateMaxTries; i++) {
-                    Script cmd = createHeartBeatCommand(primaryStoragePool, hostPrivateIp, true);
-                    result = cmd.execute();
-
-                    s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
-
-                    if (result != null) {
-                        s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; try: %s of %s.", uuid, result, i, _heartBeatUpdateMaxTries));
-                        try {
-                            Thread.sleep(_heartBeatUpdateRetrySleep);
-                        } catch (InterruptedException e) {
-                            s_logger.debug("[IGNORED] Interrupted between heartbeat retries.", e);
-                        }
-                    } else {
-                        break;
+                if (!removedPools.isEmpty()) {
+                    for (String uuid : removedPools) {
+                        removeRbdStoragePool(uuid);
                     }
-
                 }
+            }
+        }
+    }
 
-                if (result != null && rebootHostAndAlertManagementOnHeartbeatTimeout) {
-                    s_logger.warn(String.format("Write heartbeat for pool [%s] failed: %s; stopping cloudstack-agent.", uuid, result));
-                    Script cmd = createHeartBeatCommand(primaryStoragePool, null, false);
-                    result = cmd.execute();
+    private Set<String> runHeartbeatToPool(NfsStoragePool nfsStoragePool, RbdStoragePool rbdStoragePool, String uuid, Set<String> removedPools) {
+        StoragePool storage;
+        try {
+            Connect conn = LibvirtConnection.getConnection();
+            storage = conn.storagePoolLookupByUUIDString(uuid);
+            if (storage == null || storage.getInfo().state != StoragePoolState.VIR_STORAGE_POOL_RUNNING) {
+                if (storage == null) {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
+                } else {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from HA list.", uuid));
```

Comment:
not as strict on logging as @GutoVeronezi but one suggestion (making it only once)
```suggestion
                    s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from the HA list.", uuid));
```